dlk/python/dlk/templates/src/func/arch/fpga/quantized_conv2d.cpp

```diff
@@ -0,0 +1,178 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <cassert>
+#include <climits>
+#include <cmath>
+#include <cstring>
+#include <iostream>
+#include <iterator>
+
+#include "global.h"
+#include "func/apply_thresholds.h"
+#include "func/arch/quantized_conv2d_dim2col.h"
+#include "func/arch/quantized_conv2d_tiling.h"
+#include "func/arch/quantized_conv2d_kn2row.h"
+#include "func/quantized_conv2d.h"
+#include "time_measurement.h"
+
+namespace {
+
+// temporary:
+
+void func_linear_to_float(BIN_CONV_OUTPUT input[], T_INT nbit,
+                          T_FLOAT max_value, T_FLOAT output[], T_UINT in_height,
+                          T_UINT in_width, T_UINT in_depth,
+                          T_UINT in_channel = 1) {
+  T_FLOAT n = (1 << nbit) - 1;
+  unsigned num_elems = in_height * in_width * in_depth * in_channel;
+
+  for (unsigned i = 0; i < num_elems; i++) {
+    T_FLOAT tmp = (T_FLOAT)input[i];
+    tmp = tmp / n;
+    output[i] = tmp * max_value;
+  }
+}
+
+void QuantizedConv2D(QUANTIZED_NOT_PACKED input[], T_UINT kernel[],
+                            binary_convolution_parameters p) {
+  constexpr T_UINT TilingInTypeBitWidth = CHAR_BIT * sizeof(QUANTIZED_PACKED);
+  int kh = p.normal_conv_params.kernel_height;
+  int kw = p.normal_conv_params.kernel_width;
+  int padding = p.normal_conv_params.padding;
+  int ih = p.normal_conv_params.input_height;
+  int iw = p.normal_conv_params.input_width;
+  int ic = p.normal_conv_params.kernel_depth;
+  int oc = p.normal_conv_params.output_channels;
+  auto size = oc * ih * iw;
+  if (p.device_output_buf == nullptr)
+    p.device_output_buf = new BIN_CONV_OUTPUT[size]();
+  else
+    std::memset((void *)p.device_output_buf, 0, size * sizeof(BIN_CONV_OUTPUT));
+
+  if (kh == 3 && kw == 3 && padding == 1) {
+    dlk::impl::QuantizedConv2DKn2Row(input, kernel, p);
+  } else if (kh == 1 && kw == 1 && padding == 0) {
+    dlk::impl::QuantizedConv2DKn2Row(input, kernel, p);
+  } else {
+    dlk::impl::QuantizedConv2DIm2Col(input, kernel, p);
+  }
+}
+
+} // namespace
+
+void func_QuantizedConv2D(QUANTIZED_NOT_PACKED input[], T_UINT kernel[],
+                          T_FLOAT output[], T_FLOAT scaling_factor,
+                          binary_convolution_parameters p) {
+  Measurement::Start("QuantizedConv2D");
+
+  QuantizedConv2D(input, kernel, p);
+
+  Measurement::Stop();
+
+  Measurement::Start("QuantizedConv2D_ApplyScalingFactor");
+
+  unsigned out_elems = p.normal_conv_params.output_height *
+                       p.normal_conv_params.output_width *
+                       p.normal_conv_params.output_channels;
+
+  // temporary: (2^n - 1) * (max - min)
+  const T_FLOAT post_qtz_factor = 2.0f / 3.0f;
```

Comment:
@tkclimb, @antonionevado someone knows do we still need hardcoded scaling factors inside of `QuantizedConv2D()` ? Looks very hacky... :scream: